community/values/src/test/java/org/neo4j/values/storable/UTF8StringValueTest.java

```diff
@@ -59,9 +141,65 @@ public void shouldHandleOffset()
 
     private void assertSame( TextValue lhs, TextValue rhs )
     {
-        assertThat( lhs.length(), equalTo( rhs.length() ) );
+        assertThat( format( "%s.length != %s.length", lhs, rhs ), lhs.length(),
+                equalTo( rhs.length() ) );
+        assertThat( format( "%s != %s", lhs, rhs ), lhs, equalTo( rhs ) );
+        assertThat( format( "%s != %s", rhs, lhs ), rhs, equalTo( lhs ) );
+        assertThat( format( "%s.hashCode != %s.hashCode", rhs, lhs ), lhs.hashCode(), equalTo( rhs.hashCode() ) );
         assertThat( lhs, equalTo( rhs ) );
-        assertThat( rhs, equalTo( lhs ) );
-        assertThat( lhs.hashCode(), equalTo( rhs.hashCode() ) );
+    }
+
+    @Test
+    public void shouldHandleTooLargeStartPointInSubstring()
+    {
+        // Given
+        UTF8StringValue value = utf8Value( "hello".getBytes( UTF_8 ) );
+
+        // When
+        TextValue substring = value.substring( 8, 5 );
+
+        // Then
+        assertThat( substring, equalTo( StringValue.EMTPY ) );
+    }
+
+    @Test
+    public void shouldHandleTooLargeLengthInSubstring()
+    {
+        // Given
+        UTF8StringValue value = utf8Value( "hello".getBytes( UTF_8 ) );
+
+        // When
+        TextValue substring = value.substring( 3, 76 );
+
+        // Then
+        assertThat( substring.stringValue(), equalTo( "lo" ) );
+    }
+
+    @Test
+    public void shouldThrowOnNegativeStart()
+    {
+        // Given
+        UTF8StringValue value = utf8Value( "hello".getBytes( UTF_8 ) );
+
+        // Expect
+        exception.expect( IndexOutOfBoundsException.class );
+
+        // When
+        value.substring( -4, 3 );
+    }
+
+    @Test
+    public void shouldThrowOnNegativeLength()
+    {
+        // Given
+        UTF8StringValue value = utf8Value( "hello".getBytes( UTF_8 ) );
+
+        // Expect
+        exception.expect( IndexOutOfBoundsException.class );
+
+        // When
+        value.substring( 4, -3 );
     }
 }
+
```

Comment:
The style god frowns upon the double blank lines and refuses to let you compile this code.